core/src/main/java/org/bitcoinj/script/Script.java

```diff
@@ -229,38 +229,91 @@ public enum VerifyFlag {
     // Creation time of the associated keys, or null if unknown.
     @Nullable private Instant creationTime;
 
-    /** Creates an empty script that serializes to nothing. */
-    private Script() {
-        chunks = new ArrayList<>();
-    }
-
-    // Used from ScriptBuilder.
-    Script(List<ScriptChunk> chunks, Instant creationTime) {
-        this.chunks = Collections.unmodifiableList(new ArrayList<>(chunks));
-        this.creationTime = creationTime;
+    /**
+     * Construct a script that copies and wraps a given program. The array is parsed and checked for syntactic
+     * validity. Programs like this are e.g. used in {@link TransactionInput} and {@link TransactionOutput}.
+     *
+     * @param program array of program bytes
+     * @return parsed program
+     * @throws ScriptException if the program could not be parsed
+     */
+    public static Script parse(byte[] program) throws ScriptException {
+        return parse(program, TimeUtils.currentTime());
     }
 
     /**
-     * Construct a Script that copies and wraps the programBytes array. The array is parsed and checked for syntactic
-     * validity. Use this if the creation time is not known.
-     * @param programBytes Array of program bytes from a transaction.
+     * Construct a script that copies and wraps a given program, and a creation time. The array is parsed and checked
+     * for syntactic validity. Programs like this are e.g. used in {@link TransactionInput} and
+     * {@link TransactionOutput}.
+     *
+     * @param program      Array of program bytes from a transaction.
+     * @param creationTime creation time of the script
+     * @return parsed program
+     * @throws ScriptException if the program could not be parsed
      */
-    public Script(byte[] programBytes) throws ScriptException {
-        this.program = programBytes;
-        parse(programBytes);
-        this.creationTime = null;
+    public static Script parse(byte[] program, Instant creationTime) throws ScriptException {
+        Objects.requireNonNull(creationTime);
+        program = Arrays.copyOf(program, program.length); // defensive copy
+        List<ScriptChunk> chunks = new ArrayList<>(5); // common size
+        parse(program, chunks);
+        return new Script(program, chunks, creationTime);
     }
 
     /**
-     * Construct a Script that copies and wraps the programBytes array. The array is parsed and checked for syntactic
-     * validity.
-     * @param programBytes Array of program bytes from a transaction.
-     * @param creationTime creation time of the script
+     * To run a script, first we parse it which breaks it up into chunks representing pushes of data or logical
+     * opcodes. Then we can run the parsed chunks.
      */
-    public Script(byte[] programBytes, Instant creationTime) throws ScriptException {
+    private static void parse(byte[] program, List<ScriptChunk> chunks) throws ScriptException {
```

Comment:
This method should be renamed `parseChunks` (it already has a different return type than the other `static` `parse()` methods) and should allocate the `List<ScriptChunk>` internally and return it.

Reply:
I renamed the method to `parseIntoChunks()`. I could not change how the argument handling works, because two of the three usages use this method for partial parsing (in case of exception). I know this is bad, but it's out of scope of this PR to change this.